libs/parallelism/algorithms/tests/unit/container_algorithms/test_utils.hpp

```diff
@@ -40,6 +61,26 @@ namespace test {
           : base_type(base)
         {
         }
+
+        friend bool operator==(test_iterator i, Sentinel<BaseIterator> s)
+        {
+            return *i == *s;
```

Comment:
Is this correct? Why do we compare the results from dereferencing the iterator and the sentinel?

Reply:
Can you be more precise? I'm still a newbie when it comes to C++ sorry

Reply:
Sentinels can represent at least 3 different concepts: 
- iterators (the algorithm is called using a iterator pair)
- a type that encapsulates a 'sentinel value', e.g., when representing a C-style string with a pointer and the `'\0'` EOS 
- a type that encapsulates a range length, e.g. when representing a byte sequence with a pointer and a number of (consecutive) bytes

Your solution to overload the iterator's `operator==` comparing the dereferenced values will give correct answers only for the second case.

I think the correct solution would be to overload a `operator==` on the sentinel type (not the iterator) and `enable_if` it only for iterators the sentinel is valid for (`is_sentinel_for<Sent, Iter>::value == true`).

Reply:
Thanks a lot for the explanations!